src/meili/geometry_helpers.cc

```diff
@@ -0,0 +1,101 @@
+#include <valhalla/midgard/util.h>
+
+using namespace valhalla;
+using namespace valhalla::meili;
+using namespace valhalla::midgard;
+
+namespace valhalla {
+namespace meili {
+namespace helpers {
+
+// snapped point, squared distance, segment index, offset
+std::tuple<PointLL, float, typename std::vector<PointLL>::size_type, float>
+Project(const projector_t& p, Shape7Decoder<midgard::PointLL>& shape, float snap_distance) {
+  PointLL first_point(shape.pop());
+  auto closest_point = first_point;
+  auto closest_segment_point = first_point;
+  float closest_distance = p.approx.DistanceSquared(closest_segment_point);
+  size_t closest_segment = 0;
+  float closest_partial_length = 0.f;
+  float total_length = 0.f;
+
+  // for each segment
+  auto u = first_point;
+  size_t i = 0;
+  for (; !shape.empty(); ++i) {
+    // project a onto b where b is the origin vector representing this segment
+    // and a is the origin vector to the point we are projecting, (a.b/b.b)*b
+    auto v = shape.pop();
+
+    auto projection = p(u, v);
+
+    // check if this point is better
+    const auto distance = p.approx.DistanceSquared(projection);
+    if (distance < closest_distance) {
+      closest_point = std::move(projection);
+      closest_distance = distance;
+      closest_segment = i;
+      closest_partial_length = total_length;
+      closest_segment_point = u;
+    }
+
+    // total edge length
+    total_length += u.Distance(v);
+    u = v;
+  }
+
+  // percent_along is a float between 0 and 1 representing the location of
+  // the closest point on LineString to the given Point, as a fraction
+  // of total 2d line length.
+  closest_partial_length += closest_segment_point.Distance(closest_point);
+  float percent_along =
+      total_length > 0.f ? static_cast<float>(closest_partial_length / total_length) : 0.f;
+
+  // Not so much "snapping" as recognizing that floating-point has limited precision.
+  // For example:
+  // float k = 7.63513697E-9;  // valid
+  // float r = 1.f - k;        // sure why not
+  // Result: r == 1.0          // exactly 1.0?! yep.
+  //
+  // Downstream logic looks at percentages along the edge and the opp-edge. We want to be
+  // sure that values very close to 0.0 snap to 0.0 since downstream math may flip this
+  // percentage_along to (1.0-percentage_along) if the the edge is not forward, making it
+  // exactly 1.0. This causes issues because the percentage_along the opp_edge is not
+  // exactly 0.0, which is not expected.
+  //
+  // Float has "7.2" decimal digits of precision. Hence, we will consider 1e-7 as our
+  // representative small figure, below which we snap to 0.0 and within that distance
+  // of 1.0 will snap to 1.0.
+  constexpr float float_precision_at_one = 1e-7;
+  constexpr float very_close_to_one = 1.0 - float_precision_at_one;
+  if (percent_along < float_precision_at_one) {
+    percent_along = 0.0;
+  } else if (percent_along > very_close_to_one) {
+    percent_along = 1.0;
+  }
```

Comment:
other than the comments the only change is really this block yes?

Reply:
Almost.  There was this line of code previously that was occurring right before this newly added code:
`offset = std::max(0.f, std::min(offset, 1.f));`
It was basically bounding offset to [0.0,1.0].  My condition above achieves the same so I removed that previous logic.

Reply:
> i forget, is the opposing edge projected separately, ie do we call this method twice with each direction of the edge and for one direction we get a 1 where as the other direction we get a 0.00001 but not 0.0 as expected? it seems incredibly wasteful to project the same geometry twice so i dont suspect we are doing that but i didnt go check to refresh my memory. is it possible that there are more than one pair of edges at the node in question and that we get a discrepancy there about percent along as you discribed, not with an edge and is opposing edge, but with two edges that share a node but arent opposing of each other?

The issue is on this line:
https://github.com/valhalla/valhalla/blob/712824602bff7ec89f24a04c03980ecb38f9a07c/src/meili/candidate_search.cc#L84

The edge is not forward, so `1.f-percentage_along` becomes `1.f` when `percentage_along=1e-9`.

As to whether we are projecting twice, it seems to be the case. However, this PR is largely meant to fix the segfault and improve the mathematical conditioning of the algorithm.  Avoiding double projecting can be taken up in a separate PR I think.

Reply:
> Avoiding double projecting can be taken up in a separate PR I think.

yep im not suggesting you fix that in this pr if thats the case. im trying to make sure i understand correctly what the series of events is that leads to this happening